File: src/problem00682/Solution682.java
/*
 * 682. Baseball Game
 *
 * - Difficulty: Easy
 * - Topics: Array, Stack, Simulation
 */

import java.util.ArrayList;

public class Solution682 {
    public boolean isInteger(String str) {
        if (str == null) {
            return false;
        }
        try {
            Integer.parseInt(str);
        } catch (NumberFormatException nfe) {
            return false;
        }

        return true;
    }

    public int calPoints(String[] operations) {
        ArrayList<Integer> numbers = new ArrayList<>();
        for (int i = 0; i < operations.length; i++) {
            if (this.isInteger(operations[i])) {
                numbers.add(Integer.valueOf(operations[i]));
            } else {
                switch (operations[i]) {
                    case "+":
                        numbers.add(numbers.get(numbers.size() - 1)
                                + numbers.get(numbers.size() - 2));
                        break;
                    case "D":
                        numbers.add(numbers.get((numbers.size() - 1)) * 2);
                        break;
                    case "C":
                        numbers.remove(numbers.get(numbers.size() - 1));
                        break;

                }
            }
        }

        int sum = 0;
        for (int num : numbers)
            sum += num;

        return sum;
    }

    public static void main(String[] args) {
        // String[] operations = { "5", "2", "C", "D", "+" };
        // expected: 30

        // String[] operations = { "5", "-2", "4", "C", "D", "9", "+", "+" };
        // expected: 27

        String[] operations = { "1", "C" };
        // expected: 0

        long start = System.currentTimeMillis();

        Solution682 solution = new Solution682();
        System.out.println(solution.calPoints(operations));

        long end = System.currentTimeMillis();
        System.out.println("\nExecution time: " + (end - start) + " ms.");
    }
}

File: src/problem00682/solution682.js
/**
 * 682. Baseball Game
 *
 * - Difficulty: Easy
 * - Topics: Array, Stack, Simulation
 *
 * ------------------
 * Author: F. Waskito
 */

/**
 * @param {string[]} operations
 * @return {number}
 */
function calPoints(operations) {
  let numbers = [];
  for (let i = 0; i < operations.length; i++) {
    if (Number(operations[i])) {
      numbers.push(Number(operations[i]));
    } else {
      switch (operations[i]) {
        case "+":
          numbers.push(numbers.slice(-1)[0] + numbers.slice(-2)[0]);
          break;
        case "D":
          numbers.push(numbers.slice(-1)[0] * 2);
          break;
        case "C":
          numbers.pop();
          break;
      }
    }
  }

  let sum = 0;
  numbers.forEach((x) => {
    sum += x;
  });

  return sum;
}

// operations = ["5", "2", "C", "D", "+"];
// expected: 30

// let operations = ["5", "-2", "4", "C", "D", "9", "+", "+"];
// expected: 27

let operations = ["1", "C"];
// expected: 0

const start = performance.now();

console.log(calPoints(operations));

const end = performance.now();
console.log(`\nExecution time: ${end - start} ms.`);
